clients/da-vinci-client/src/main/java/com/linkedin/davinci/store/view/ChangeCaptureViewWriter.java

```diff
@@ -0,0 +1,196 @@
+import com.linkedin.venice.exceptions.VeniceException;
+import com.linkedin.venice.kafka.protocol.ControlMessage;
+import com.linkedin.venice.kafka.protocol.VersionSwap;
+import com.linkedin.venice.kafka.protocol.enums.ControlMessageType;
+import com.linkedin.venice.meta.PartitionerConfig;
+import com.linkedin.venice.meta.Store;
+import com.linkedin.venice.meta.Version;
+import com.linkedin.venice.partitioner.VenicePartitioner;
+import com.linkedin.venice.schema.rmd.RmdUtils;
+import com.linkedin.venice.serialization.VeniceKafkaSerializer;
+import com.linkedin.venice.serialization.avro.VeniceAvroKafkaSerializer;
+import com.linkedin.venice.utils.PartitionUtils;
+import com.linkedin.venice.utils.VeniceProperties;
+import com.linkedin.venice.views.ChangeCaptureView;
+import com.linkedin.venice.writer.VeniceWriter;
+import com.linkedin.venice.writer.VeniceWriterFactory;
+import com.linkedin.venice.writer.VeniceWriterOptions;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+public class ChangeCaptureViewWriter extends VeniceViewWriter {
+  private static final Logger LOGGER = LogManager.getLogger(ChangeCaptureViewWriter.class);
+  final private ChangeCaptureView internalView;
+  private VeniceWriter veniceWriter;
+  private final Object2IntMap<String> kafkaClusterUrlToIdMap;
+
+  public ChangeCaptureViewWriter(
+      VeniceConfigLoader props,
+      Store store,
+      Schema keySchema,
+      Map<String, String> extraViewParameters) {
+    super(props, store, keySchema, extraViewParameters);
+    internalView = new ChangeCaptureView(props.getCombinedProperties().toProperties(), store, extraViewParameters);
+    kafkaClusterUrlToIdMap = props.getVeniceServerConfig().getKafkaClusterUrlToIdMap();
+  }
+
+  @Override
+  public void processRecord(
+      ByteBuffer newValue,
+      ByteBuffer oldValue,
+      ByteBuffer key,
+      int version,
+      int newValueSchemaId,
+      int oldValueSchemaId,
+      GenericRecord replicationMetadataRecord) {
+    // TODO: not sold about having currentValue in the interface but it VASTLY simplifies a lot of things with regards
+    // to dealing with compression/chunking/etc. in the storage layer.
+
+    RecordChangeEvent recordChangeEvent = new RecordChangeEvent();
+    recordChangeEvent.currentValue = constructValueBytes(newValue, newValueSchemaId);
+    recordChangeEvent.previousValue = constructValueBytes(oldValue, oldValueSchemaId);
+    recordChangeEvent.key = key;
+    recordChangeEvent.replicationCheckpointVector = RmdUtils.extractOffsetVectorFromRmd(replicationMetadataRecord);
+
+    if (veniceWriter == null) {
+      initializeVeniceWriter(version);
```

Comment:
The `processRecord` here is called from `processMessageAndMaybeProduceToKafka` in consumer thread. There might be race conditions here to create multiple `VeniceWriter`. Just wondering, since there is 1:1 mapping from `ChangeCaptureViewWriter` to `VeniceWriter`, is it OK to create it in constructor?

Reply:
Function is synchronized, but based on our current discussion: will add a TODO to consider a shared producer for each viewType.